### PR TITLE
Batch of fixes and improvements (+some refactor of things thrown into it)

### LINKED ISSUE #329: [Issue] Possible Resolution Problem for use in Full Screen
So I have found on my own that the only way to get the Command Console GUI to function normally in Full Screen and let me select exactly where I am clicking is to make sure that both the game resolution and my monitor resolution are the same (ie. set both to 1920 x 1080).

I am wondering if maybe this is because the selectable fields and the graphical interface are based on different resolutions (one is based on game resolution, the other is based on set monitor resolution)?

Thought I would share this, as this is the only way I have found to get the Command Console to work in Full Screen.

### PR DESCRIPTION
**Pull request by WSSDude420**
Fixes #329, #319, #286, #283, #275, #270, #260, #257, #236, #228, #193, #186, #137

Following text is from initial PR, does not contain all changes!

> Main one here being more reliable resize hook for Windows 10 and removal of less reliable one pulled before.
> 
> It seems it was not as bulletproof as first thought... So even though I wanted to left it in as partial fix at least for Windows 7, I rather removed it.
> 
> Other major change is complete refactor of kiero, as we already made it our own I'd say.
> I also expanded its method table and would like to add some more logic into methods table (prolly create some massive enum, so we it can be more easily seen to what we try to hook right from code... plain numbers suck balls).

### CODE REVIEW

**Comment by WSSDude420** [comment]
xmake keeps failling @yamashi dunno why exactly...

I just tried clean checkout of that branch, updated repos, rebuild,...
All works fine, something seems to be messed up badly with this check, it cannot even get to the build it seems

**Comment by WSSDude420** [comment]
Ok, so from #347, check fail is probably some error regarding xmake itself, lua package related one it seems...

**Comment by yamashi** [comment]
Lua.org's cert expired...

**Comment by WSSDude420** [comment]
Well, that explains it I guess... Hope it will be resolved soon...

**Comment by WSSDude420** [comment]
xmake issue resolved!
https://github.com/yamashi/CyberEngineTweaks/issues/347#issuecomment-752918170

File: xmake.lua [hunk 1]
```diff
@@ -1,4 +1,4 @@
-set_languages("cxx17")
+set_languages("cxx20")
```
Comment (yamashi) [inline xmake.lua:1]: This breaks visual studio projects afaik

Comment (WSSDude420) [reply xmake.lua:1]: Solution was created with no issues with this set as is here, just checked by following steps in BUILDING.md.

It resolves compiler warning, that is why I changed it.
As said, just tested that it works. Here is output from build without this and after that with it.
All is fine and dandy :)

Output without change from building generated solution:
```
Build started...
1>------ Build started: Project: RED4ext.SDK, Configuration: debug x64 ------
2>------ Build started: Project: cyber_engine_tweaks, Configuration: debug x64 ------
2>$xmake config -P . -p windows -m debug -a x64 -o "build"
1>$xmake config -P . -p windows -m debug -a x64 -o "build"
2>checking for Microsoft Visual Studio (x64) version ... 2019
1>$xmake build -P . -w RED4ext.SDK
2>$xmake build -P . -w cyber_engine_tweaks
1>[ 18%]: compiling.debug vendor\RED4ext.SDK\src\CName.cpp
1>[ 72%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Stack.cpp
1>[ 45%]: compiling.debug vendor\RED4ext.SDK\src\RTTISystem.cpp
1>[ 81%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Utils.cpp
1>[ 27%]: compiling.debug vendor\RED4ext.SDK\src\CString.cpp
1>[ 36%]: compiling.debug vendor\RED4ext.SDK\src\GameEngine.cpp
1>[ 54%]: compiling.debug vendor\RED4ext.SDK\src\RTTITypes.cpp
1>[ 63%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Functions.cpp
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
1>[ 90%]: archiving.debug RED4ext.SDK.lib
1>[100%]: build ok!
1>Done building project "RED4ext.SDK.vcxproj".
2>[  4%]: compiling.debug src\stdafx.h
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>[ 24%]: compiling.debug src\dllmain.cpp
2>[ 26%]: compiling.debug src\Image.cpp
2>[ 28%]: compiling.debug src\kiero\kiero.cpp
2>[ 31%]: compiling.debug src\Options.cpp
2>[ 33%]: compiling.debug src\overlay\imgui_impl_dx12.cpp
2>[ 35%]: compiling.debug src\overlay\imgui_impl_win32.cpp
2>[ 37%]: compiling.debug src\overlay\Overlay.cpp
2>[ 40%]: compiling.debug src\overlay\Overlay_D3D12.cpp
2>[ 42%]: compiling.debug src\overlay\Overlay_Hooks.cpp
2>[ 44%]: compiling.debug src\patches\Avx.cpp
2>[ 46%]: compiling.debug src\patches\DisableBoundaries.cpp
2>[ 48%]: compiling.debug src\patches\DisableIntroMovies.cpp
2>[ 51%]: compiling.debug src\patches\DisableVignette.cpp
2>[ 53%]: compiling.debug src\patches\EnableDebug.cpp
2>[ 55%]: compiling.debug src\patches\OptionsPatch.cpp
2>[ 57%]: compiling.debug src\patches\RemovePeds.cpp
2>[ 60%]: compiling.debug src\patches\SkipStartScreen.cpp
2>[ 62%]: compiling.debug src\patches\Smt.cpp
2>[ 64%]: compiling.debug src\patches\VirtualInput.cpp
2>[ 66%]: compiling.debug src\Pattern.cpp
2>[ 68%]: compiling.debug src\reverse\BasicTypes.cpp
2>[ 71%]: compiling.debug src\reverse\Converter.cpp
2>[ 73%]: compiling.debug src\reverse\SingletonReference.cpp
2>[ 75%]: compiling.debug src\reverse\StrongReference.cpp
2>[ 77%]: compiling.debug src\reverse\Type.cpp
2>[ 80%]: compiling.debug src\reverse\WeakReference.cpp
2>[ 82%]: compiling.debug src\scripting\GameOptions.cpp
2>[ 84%]: compiling.debug src\scripting\ScriptContext.cpp
2>[ 86%]: compiling.debug src\scripting\Scripting.cpp
2>[ 88%]: compiling.debug src\scripting\ScriptStore.cpp
2>[ 91%]: compiling.debug src\Utils.cpp
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>cl : Command line warning D9025: overriding '/std:c++17' with '/std:c++latest'
2>[ 97%]: linking.debug cyber_engine_tweaks.asi
2>[100%]: build ok!
2>Done building project "cyber_engine_tweaks.vcxproj".
========== Build: 2 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

Output without change from building generated solution:
```
Rebuild started...
1>------ Rebuild All started: Project: RED4ext.SDK, Configuration: debug x64 ------
2>------ Rebuild All started: Project: cyber_engine_tweaks, Configuration: debug x64 ------
1>$xmake clean -P . -a RED4ext.SDK
2>$xmake clean -P . -a cyber_engine_tweaks
1>$xmake config -P . -p windows -m debug -a x64 -o "build"
2>checking for architecture ... x64
2>checking for Microsoft Visual Studio (x64) version ... 2019
2>$xmake config -P . -p windows -m debug -a x64 -o "build"
1>checking for Microsoft Visual Studio (x64) version ... 2019
2>$xmake build -P . -w cyber_engine_tweaks
1>$xmake build -P . -w RED4ext.SDK
2>[  4%]: compiling.debug src\stdafx.h
2>[  6%]: compiling.debug vendor\RED4ext.SDK\src\CName.cpp
2>[  8%]: compiling.debug vendor\RED4ext.SDK\src\CString.cpp
2>[ 11%]: compiling.debug vendor\RED4ext.SDK\src\GameEngine.cpp
2>[ 13%]: compiling.debug vendor\RED4ext.SDK\src\RTTISystem.cpp
2>[ 15%]: compiling.debug vendor\RED4ext.SDK\src\RTTITypes.cpp
2>[ 17%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Functions.cpp
2>[ 20%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Stack.cpp
2>[ 22%]: compiling.debug vendor\RED4ext.SDK\src\Scripting\Utils.cpp
2>[ 24%]: compiling.debug src\dllmain.cpp
2>[ 26%]: compiling.debug src\Image.cpp
2>[ 28%]: compiling.debug src\kiero\kiero.cpp
2>[ 31%]: compiling.debug src\Options.cpp
2>[ 33%]: compiling.debug src\overlay\imgui_impl_dx12.cpp
2>[ 35%]: compiling.debug src\overlay\imgui_impl_win32.cpp
2>[ 37%]: compiling.debug src\overlay\Overlay.cpp
2>[ 40%]: compiling.debug src\overlay\Overlay_D3D12.cpp
2>[ 42%]: compiling.debug src\overlay\Overlay_Hooks.cpp
2>[ 44%]: compiling.debug src\patches\Avx.cpp
2>[ 46%]: compiling.debug src\patches\DisableBoundaries.cpp
2>[ 48%]: compiling.debug src\patches\DisableIntroMovies.cpp
2>[ 51%]: compiling.debug src\patches\DisableVignette.cpp
2>[ 53%]: compiling.debug src\patches\EnableDebug.cpp
2>[ 55%]: compiling.debug src\patches\OptionsPatch.cpp
2>[ 57%]: compiling.debug src\patches\RemovePeds.cpp
2>[ 60%]: compiling.debug src\patches\SkipStartScreen.cpp
2>[ 62%]: compiling.debug src\patches\Smt.cpp
2>[ 64%]: compiling.debug src\patches\VirtualInput.cpp
2>[ 66%]: compiling.debug src\Pattern.cpp
2>[ 68%]: compiling.debug src\reverse\BasicTypes.cpp
2>[ 71%]: compiling.debug src\reverse\Converter.cpp
2>[ 73%]: compiling.debug src\reverse\SingletonReference.cpp
2>[ 75%]: compiling.debug src\reverse\StrongReference.cpp
2>[ 77%]: compiling.debug src\reverse\Type.cpp
2>[ 80%]: compiling.debug src\reverse\WeakReference.cpp
2>[ 82%]: compiling.debug src\scripting\GameOptions.cpp
2>[ 84%]: compiling.debug src\scripting\ScriptContext.cpp
2>[ 86%]: compiling.debug src\scripting\Scripting.cpp
2>[ 88%]: compiling.debug src\scripting\ScriptStore.cpp
2>[ 91%]: compiling.debug src\Utils.cpp
2>[ 93%]: archiving.debug RED4ext.SDK.lib
2>[ 97%]: linking.debug cyber_engine_tweaks.asi
2>[100%]: build ok!
1>[100%]: build ok!
========== Rebuild All: 2 succeeded, 0 failed, 0 skipped ==========
```

**Comment by WSSDude420** [comment]
Fixes #329, #319, #286, #283, #275, #270, #260, #257, #236, #228, #193, #186, #137

File: src/Options.cpp [hunk 2]
```diff
@@ -107,6 +110,9 @@ Options::Options(HMODULE aModule)
         this->ConsoleChar = MapVirtualKeyA(this->ConsoleKey, MAPVK_VK_TO_CHAR);
     }
 
+    std::string scriptsPath = this->ScriptsPath.string();
+    spdlog::info("LUA scripts search path: \"{}\"", scriptsPath);
```
Comment (yamashi) [inline src/Options.cpp:114]: LUA = Lua

File: src/overlay/Overlay_D3D12.cpp [hunk 3]
```diff
@@ -87,32 +89,45 @@ bool Overlay::InitializeD3D12(IDXGISwapChain3* pSwapChain)
 
     if (m_initialized) 
     {
-        if (m_pdxgiSwapChain != pSwapChain)
+        CComPtr<IDXGISwapChain3> pSwapChain3;
+		if (FAILED(pSwapChain->QueryInterface(IID_PPV_ARGS(&pSwapChain3))))
+		{
+            spdlog::error("Overlay::InitializeD3D12() - unable to query pSwapChain interface for IDXGISwapChain3! (pSwapChain = {0})", reinterpret_cast<void*>(pSwapChain));
```
Comment (yamashi) [inline src/overlay/Overlay_D3D12.cpp:95]: Indentation is weird

File: xmake.lua (hunk 1 above)
Comment (yamashi) [reply xmake.lua:1]: Yes but intelisense doesn't work with C++20 yet

File: src/scripting/Scripting.cpp [hunk 4]
```diff
@@ -396,7 +396,18 @@ void Scripting::Initialize()
         Overlay::Get().Log(oss.str());
     };
 
-    m_lua.do_file((Options::Get().Path / "scripts" / "autoexec.lua").string());
+
+    // execute autoexec.lua inside our default script directory
+    std::filesystem::path defPath = Options::Get().Path / "scripts";
```
Comment (yamashi) [inline src/scripting/Scripting.cpp:401]: This is the same path as ScriptsPath, I don't understand what this is supposed to accomplish?

**Review by yamashi — CHANGES_REQUESTED** [review]

**Comment by WSSDude420** [comment]
Ok, I'll stop pushing now...

**Comment by WSSDude420** [comment]
I'm not gonna fight that xmake.lua change... Can make it for myself so here you go.

Although, tbh, there is really nothing bad with it. Solutions work, intellisense works, everything works with it and we do not have tons of warnings with it...

But whatever, as you wish, reverted it. Still don't see much sense in keeping it the way it is though.

**Comment by WSSDude420** [comment]
User-configurable Lua path testing against default path was corrected, there were few issues.

Also removed redundant backslashes in paths.

I think all things here were tested. I will make another PR for next changes, I personally consider this closed.

If you have some more issues with this you would like changed, just let me know :)

File: src/overlay/Overlay_D3D12.cpp [hunk 5]
```diff
@@ -87,32 +89,45 @@ bool Overlay::InitializeD3D12(IDXGISwapChain3* pSwapChain)
 
     if (m_initialized) 
     {
-        if (m_pdxgiSwapChain != pSwapChain)
+        CComPtr<IDXGISwapChain3> pSwapChain3;
+		if (FAILED(pSwapChain->QueryInterface(IID_PPV_ARGS(&pSwapChain3))))
```
Comment (yamashi) [inline src/overlay/Overlay_D3D12.cpp:93]: weird indent

**Review by yamashi — CHANGES_REQUESTED** [review]

**Review by yamashi — APPROVED** [review]

**Pull request merged**